app/controllers/api/v1/users/sessions_controller.rb

```diff
@@ -0,0 +1,40 @@
+# frozen_string_literal: true
+
+module Api
+  module V1
+    module Users
+      class SessionsController < Devise::SessionsController
+        include ValidClientConcern
+
+        respond_to :json
+
+        def create
+          user = User.find_for_database_authentication(email: user_params[:email])
+          valid_password = user&.valid_password?(user_params[:password])
+
+          return render_error([I18n.t('devise.failure.invalid', authentication_keys: :email)]) unless valid_password
```

Comment:
This code is trying to be too smart IMO 😅 Prefer simpler:

```
user = User.find_by(email: user_params[:email])

if user&.valid_password?(user_params[:password])
  #...
else
  #...
end
```

Do not over-abuse `unless` 🧑‍🏫